plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse_token.go

```diff
@@ -0,0 +1,138 @@
+// After successfull authentication a token will be received, that can be used in all subsequent calls for
+// authorization.
+// This file encapsulates for requesting this token and for checking if it is still valid or might be outdated.
+// The implementation is based on the 'ctrlX CORE - Authorization and Authentication API'
+// The API description can be found at:
+// https://boschrexroth.github.io/rest-api-description/ctrlx-automation/ctrlx-core/
+
+const (
+	validateTokenPath = "/identity-manager/api/v2/auth/token/validity"
+	identityPath      = "/identity-manager/api/v2/auth/token"
+)
+
+// token represents the json structure that will be returned from the device after successfull authentication.
+// The token can be used for further API calls and needs to be encoded in the 'authorization' field of the
+// http header.
+type token struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+}
+
+// validToken respresents the json structure that will be returned by the device
+// after a request to check if the given token is still valid.
+type validToken struct {
+	Valid bool `json:"valid"`
+}
+
+// String builds the token string which is used in the 'authorization' field of the http header.
+func (t *token) String() string {
+	return t.TokenType + " " + t.AccessToken
+}
+
+type tokenManager struct {
+	url        string
+	username   string
+	password   string
+	connection *http.Client
+	log        telegraf.Logger
+
+	mux   sync.Mutex // mux used to serialize authorization requests
+	token token
+}
+
+// checkAuthToken checks if the tokenManager has valid authentication token.
+func (m *tokenManager) checkAuthToken() (isValid bool, err error) {
+	if m.token.AccessToken == "" {
+		return false, nil
+	}
+
+	u := m.url + validateTokenPath
+
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return false, fmt.Errorf("failed to create request to validate token: %v", err)
+	}
+
+	req.Header.Add("Authorization", m.token.String())
+
+	resp, err := m.connection.Do(req)
+	if err != nil {
+		return false, fmt.Errorf("failed to execute request to validate token: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("failed to read response to validate token: %v", err)
+	}
+	var vt validToken
+	err = json.Unmarshal(body, &vt)
+	if err != nil {
+		return false, fmt.Errorf("failed to unmarshal validate token response: %v", err)
+	}
+	return vt.Valid, nil
+}
+
+// requestAuthToken gets a valid authentication token from the ctrlX CORE using username and password.
+// The token is necessary to create a subscription. An error will be returned, when no token could get aquired.
+func (m *tokenManager) requestAuthToken() (token, error) {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	v, err := m.checkAuthToken()
+	if err != nil {
+		// Something went wrong during token validation
+		m.log.Error(err)
```

Comment:
In this case you want to renew the token? If so, maybe say so in a comment to ensure the reader that this is the intended behavior.

Reply:
Done